File: next/internal/monitor/targets.go
package monitor

import (
	"crypto/sha1"
	"encoding/hex"
	"net/url"
	"strings"

	"netwatcher/next/internal/domain"
)

func targetID(value string) string {
	sum := sha1.Sum([]byte(strings.ToLower(strings.TrimSpace(value))))
	return hex.EncodeToString(sum[:6])
}

func DefaultTargets() []domain.Target {
	targets := make([]domain.Target, 0, 3)
	if gateway := detectDefaultGateway(); gateway != "" {
		targets = append(targets, domain.Target{ID: targetID("gateway:" + gateway), Name: "Default Gateway", Host: gateway, Kind: "local", Mode: "ping"})
	}
	targets = append(targets,
		domain.Target{ID: targetID("cloudflare:1.1.1.1"), Name: "Cloudflare", Host: "1.1.1.1", Kind: "internet", Mode: "ping"},
		domain.Target{ID: targetID("google:8.8.8.8"), Name: "Google", Host: "8.8.8.8", Kind: "internet", Mode: "ping"},
	)
	return targets
}

func ParseTarget(raw string) domain.Target {
	raw = strings.TrimSpace(raw)
	t := domain.Target{ID: targetID(raw), Name: "Custom: " + raw, Host: raw, Kind: "internet", Mode: "ping", Custom: true}
	lower := strings.ToLower(raw)
	switch {
	case strings.HasPrefix(lower, "tcp://"):
		host := strings.TrimSpace(raw[len("tcp://"):])
		t.Host = host
		t.Name = "TCP: " + host
		t.Mode = "tcp"
	case strings.HasPrefix(lower, "http://") || strings.HasPrefix(lower, "https://"):
		if parsed, err := url.Parse(raw); err == nil && parsed.Host != "" {
			t.Name = strings.ToUpper(parsed.Scheme) + ": " + parsed.Host
			t.Mode = strings.ToLower(parsed.Scheme)
		}
	}
	return t
}

func TargetsFromConfig(custom []string) []domain.Target {
	targets := DefaultTargets()
	seen := map[string]bool{}
	for _, t := range targets {
		seen[t.ID] = true
	}
	for _, raw := range custom {
		t := ParseTarget(raw)
		if strings.TrimSpace(t.Host) == "" || seen[t.ID] {
			continue
		}
		seen[t.ID] = true
		targets = append(targets, t)
	}
	return targets
}

File: rust_core/src/targets.rs
use sha1::{Digest, Sha1};
use url::Url;

use crate::models::Target;

fn target_id(value: &str) -> String {
    let mut hasher = Sha1::new();
    hasher.update(value.trim().to_ascii_lowercase().as_bytes());
    let output = hasher.finalize();
    output[..6].iter().map(|byte| format!("{byte:02x}")).collect()
}

pub fn default_targets() -> Vec<Target> {
    let mut targets = Vec::new();
    if let Some(gateway) = detect_default_gateway() {
        targets.push(Target {
            id: target_id(&format!("gateway:{gateway}")),
            name: "Default Gateway".into(),
            host: gateway,
            kind: "local".into(),
            mode: "ping".into(),
            custom: false,
        });
    }
    targets.extend([
        Target {
            id: target_id("cloudflare:1.1.1.1"),
            name: "Cloudflare".into(),
            host: "1.1.1.1".into(),
            kind: "internet".into(),
            mode: "ping".into(),
            custom: false,
        },
        Target {
            id: target_id("google:8.8.8.8"),
            name: "Google".into(),
            host: "8.8.8.8".into(),
            kind: "internet".into(),
            mode: "ping".into(),
            custom: false,
        },
    ]);
    targets
}

pub fn parse_target(raw: &str) -> Target {
    let raw = raw.trim();
    let mut target = Target {
        id: target_id(raw),
        name: format!("Custom: {raw}"),
        host: raw.into(),
        kind: "internet".into(),
        mode: "ping".into(),
        custom: true,
    };
    let lower = raw.to_ascii_lowercase();
    if lower.starts_with("tcp://") {
        let host = raw[6..].trim();
        target.host = host.into();
        target.name = format!("TCP: {host}");
        target.mode = "tcp".into();
    } else if lower.starts_with("http://") || lower.starts_with("https://") {
        if let Ok(url) = Url::parse(raw) {
            if let Some(host) = url.host_str() {
                target.name = format!("{}: {host}", url.scheme().to_ascii_uppercase());
                target.mode = url.scheme().to_ascii_lowercase();
            }
        }
    }
    target
}

pub fn all_targets(custom: &[String]) -> Vec<Target> {
    let mut values = default_targets();
    for raw in custom {
        let target = parse_target(raw);
        if target.host.trim().is_empty()
            || values.iter().any(|existing| existing.id == target.id)
        {
            continue;
        }
        values.push(target);
    }
    values
}

fn detect_default_gateway() -> Option<String> {
    #[cfg(windows)]
    {
        use std::process::Command;
        let output = Command::new("route")
            .args(["print", "-4", "0.0.0.0"])
            .output()
            .ok()?;
        let text = String::from_utf8_lossy(&output.stdout);
        for line in text.lines() {
            let columns: Vec<_> = line.split_whitespace().collect();
            if columns.len() >= 4 && columns[0] == "0.0.0.0" && columns[1] == "0.0.0.0" {
                return Some(columns[2].to_string());
            }
        }
    }
    None
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_supported_custom_target_formats() {
        assert_eq!(parse_target("1.1.1.1").mode, "ping");
        assert_eq!(parse_target("tcp://example.com:443").mode, "tcp");
        assert_eq!(parse_target("https://example.com/test").mode, "https");
    }
}
